### PR TITLE
EZP-25926 - Ability to update user login value

### PR DESCRIPTION
**Pull request by alexwilson**
Attempting [EZP-25926](https://jira.ez.no/browse/EZP-25926)!

Have implemented the new option with both a passing and a deliberately failing test.

Whilst changing validation rules, on @bdunogier's advice I moved that logic to a number of private functions.  Doing that, I realised that the same logic was being used for the CreateUserStruct, so I changed that to use the same logic which I feel has helped cut down on some cruft.

### CODE REVIEW

**Comment by antoligy** [comment]
For your approval @bdunogier @andrerom!

The two failing tests are what appear to be an assertion on a DateTime (@danzabar's favourite type of failed test!) and an unrelated Behat issue with the QueryController.

**Comment by antoligy** [comment]
Something that has come up as a topic of discussion is if we can expose these new private methods as a means of validating information in a Symfony form, or in other places - This allows for more granular validation without having to deconstruct an `InvalidArgumentException`.

HOWEVER, this would also be an API change, and not one I'm happy about.  Something for a future PR, perhaps?

**Comment by bdunogier** [comment]
Quick answer without reading the code: "Exposing private methods" ➡️  "refactor to external services".

**Comment by antoligy** [comment]
I was thinking the same thing, would you think that should be something along the lines of a helper (i.e. like `ezpublish.field_helper`) or something else?  I'm becoming [less and less of a fan of the former option](http://www.yegor256.com/2015/03/09/objects-end-with-er.html).

Also should that be a part of the same PR, or should I save that for after this has been completed?

File: eZ/Publish/Core/Repository/UserService.php [hunk 1]
```diff
@@ -1195,4 +1193,84 @@ protected function getSortClauseBySortField($sortField, $sortOrder = Location::S
                 return new SortClause\Location\Path($sortOrder);
         }
     }
+
+    /**
+     * Given a login, determines if it is suitable for use.
+     *
+     * @param  string  $login
```
Comment (bdunogier) [inline eZ/Publish/Core/Repository/UserService.php]: CS: extra spaces in phpdoc. Valid for the ones below as well.

**Comment by bdunogier** [comment]
The approach looks okay to me. What do you think, @andrerom / @pspanja ?

File: eZ/Publish/Core/Repository/UserService.php [hunk 2]
```diff
@@ -695,25 +685,32 @@ public function updateUser(APIUser $user, UserUpdateStruct $userUpdateStruct)
...
+        if ($passwordMissing || $passwordInvalid) {
             throw new InvalidArgumentValue('password', $userUpdateStruct->password, 'UserUpdateStruct');
         }
 
-        if ($userUpdateStruct->enabled !== null && !is_bool($userUpdateStruct->enabled)) {
+        if ($userUpdateStruct->email !== null && !$this->isValidEmail($userUpdateStruct->email)) {
+            throw new InvalidArgumentValue('email', $userUpdateStruct->email, 'UserUpdateStruct');
+        }
+
+        if ($userUpdateStruct->login !== null && !$this->isValidLogin($userUpdateStruct->login)) {
```
Comment (andrerom) [inline eZ/Publish/Core/Repository/UserService.php]: what if set to same value as from before?

File: eZ/Publish/Core/Repository/UserService.php [hunk 3]
```diff
@@ -601,11 +590,12 @@ public function loadUserByCredentials($login, $password)
...
      * @return \eZ\Publish\API\Repository\Values\User\User
      *
-     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException if a user with the given credentials was not found
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException        if a user with the given credentials was not found
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException if an invalid login is specified
      */
     public function loadUserByLogin($login)
     {
-        if (!is_string($login) || empty($login)) {
+        if (!$this->isValidString($login)) {
```
Comment (andrerom) [inline eZ/Publish/Core/Repository/UserService.php]: not relevant change

**Comment by andrerom** [comment]
would need to reduce the diff here, patch is doing more then what issue is about.

besides that I'm not against, can't remember why updating login was left out _(pretty sure there was a reason)_, maybe @pspanja remembers or have clues.

**Comment by antoligy** [comment]
@andrerom I extended the scope of the diff, so as to avoid having duplicate validation on the similar API methods, though happy to remove if this is the consensus.

Will address the other issues, thanks for looking @andrerom @bdunogier

File: eZ/Publish/Core/Repository/UserService.php (hunk 2 above)
Comment (antoligy) [inline eZ/Publish/Core/Repository/UserService.php]: Nice catch - I guess `isValidLogin` would raise an exception - I'll a test for this scenario, first thoughts are that if login matches the loadedUser's login then it satisfies this condition?

Comment (andrerom) [inline eZ/Publish/Core/Repository/UserService.php]: if user id is the same, then it should not return false, see content service in terms of remote id updates.

**Comment by pspanja** [comment]
> maybe pspanja remembers or have clues.

I don't, but let's try pinging @emodric as he worked on this part :)

**Comment by emodric** [comment]
> I don't, but let's try pinging @emodric as he worked on this part :)

I don't remember a reason. We simply followed what was in the original legacy API.

But I think the reason is that changing username is not really trivial though:

1) This code requires that password is supplied when changing the login. While this is certainly legit, it is not in any way 100% correct, especially when we introduce bcrypt hashing later on.

2) What if I want to change the user's login without changing the password? (since, again, password is not related to username in half a cases) This is not possible with this patch unless the user is the one initiating the username change (i.e. knowing the current password). But in that case, this still lacks checks whether the current password is correct before changing the username. 

Considering the two points above, I don't really think this is a good idea, at least not in this form.

I'm definitely for reducing the diff though, it can be tackled in another PR, together with replacing `ezcMailTools` with `egulias/email-validator` maybe?

**Comment by andrerom** [comment]
> BTW, ezcMailTools could probably be replaced with something like egulias/email-validator? :)

true, but separate PR please, that would imply removing the remaining dependency on Zeta stuff.

**Comment by antoligy** [comment]
I'll raise something in JIRA about replacing the email address validator.

**Comment by antoligy** [comment]
About requiring or not requiring the username based on what the hash type is, I don't see why we can't extend this logic to introduce conditional validation based on whether or not the hash type requires it (or other fields!).  That logic would almost certainly have to live outside of this particular method as it's too complex to reliably repeat.

**Comment by bdunogier** [comment]
Should we close this given the lack of movement ? I'm not against the change by itself, even though the "changing the login may require that you change the password as well" part sounds a bit complicated.

**Comment by antoligy** [comment]
Apologies, I let this slip!  I'd still like to see this happen, but am not sure entirely how to proceed.

I'm not sure that there's any good way around the field requirement.

Is reducing the diff the main thing I should action here?  Potentially some of this cleanup can be done in subsequent PRs instead of here!

**Comment by andrerom** [comment]
> Is reducing the diff the main thing I should action here? Potentially some of this cleanup can be done in subsequent PRs instead of here!

yes, thats the remaining thing, we try to make sure PR's handle one topic, and we make sure to try to get it right first before merging, cause once it's merged it's used _and_ people tend to move on.

**Pull request closed without merging**